13 tweak links (#64)
pharmaverse/admiralci: vignettes/tweak-page-hooks.Rmd

---
title: "Custom pkgdown tweak page hooks"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Custom pkgdown tweak page hooks}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

This article will go over the following:

> * The process for using custom hooks in R packages

Reviewer: Is there any documentation around hooks - I have never used them before but now curious.

* Example how to create custom hooks

> For more information about hooks, please read `help("getHook", "base")`.

Reviewer: Is it possible to put in the issues this is addressing for admiral?  My understanding is this is helping to solve the absolute link issue.  But it would be nice to either include the issues this solves for future people reading this vignette and them trying to understand why this was made????

## Custom hooks

> Currently, the pkgdown do not allow set up any custom action, the changes apply during the executing will modify the `tweak_page` function to execute extra code base on the hook content.

Reviewer: ```suggestion
Currently,  `{pkgdown}` does not allow the set up of any custom action, the changes apply during the executing will modify the `tweak_page` function to execute extra code based on the hook content.
```

Reviewer: I tried to wordsmith this a bit, but I'm not familiar with what we are doing here honestly.

To reproduce this behavior on local environment:

```r
pkgdown_env <- asNamespace("pkgdown")
rlang::env_unlock(env = pkgdown_env)
rlang::env_binding_unlock(env = pkgdown_env)
# Modify tweak_page
pkgdown_env$call_hook <- function(hook_name, ...) {
  # Get hooks from base::getHook
  hooks <- getHook(paste0("UserHook::admiralci::", hook_name))
  if (!is.list(hooks)) {
    hooks <- list(hooks)
  }

  # Evaluate hooks
  purrr::map(hooks, function(fun) {
    fun(...)
  }) %>%
    invisible()
}
environment(pkgdown_env$call_hook) <- pkgdown_env

tweak_page <- body(pkgdown_env$tweak_page)
body(pkgdown_env$tweak_page) <-
  as.call(
    append(
      as.list(tweak_page),
      expression(call_hook("tweak_page", html, name, pkg)),
      after=length(tweak_page)
    )
  )

rlang::env_binding_lock(env = pkgdown_env)
rlang::env_lock(pkgdown_env)
```

This part is executed automatically on GitHub CI. The developer need to define only the hooks.

## Creating hooks

> The `UserHook::admiralci::tweak_page` hook should be defined as function. Please note that an error in function will stop the pkgdown from build. The below code showing an example of `tweak_page` function. The most convenient way to set up the hook is to define them in the package `.onLoad` event. 

Reviewer: ```suggestion
The `UserHook::admiralci::tweak_page` hook should be defined as function. Please note that an error in function will stop the pkgdown from building the site. The below code showing an example of `tweak_page` function. The most convenient way to set up the hook is to define them in the package `.onLoad` event. 
```
